framework/src/parser/Parser.C

```diff
@@ -139,425 +238,320 @@ Parser::getFileName(bool stripLeadingPath) const
-{
-  if (!_sections_read && error_on_warn)
-    // The user has requested errors but we haven't done any parsing yet so throw an error
-    mooseError("No parsing has been done, so checking for overridden parameters is not possible");
-
-  std::set<std::string> overridden_vars = _getpot_file_error_checking.get_overridden_variables();
-
-  if (!overridden_vars.empty())
+  catch (hit::ParseError & err)
   {
-    std::ostringstream oss;
-
-    oss << "The following variables were overridden or supplied multiple times:\n";
-    for (const auto & name : overridden_vars)
-      oss << name << "\n";
+    mooseError(err.what());
+  }
 
-    if (error_on_warn)
-      mooseError(oss.str());
-    else
-      mooseWarning(oss.str());
+  // expand ${bla} parameter values and mark/include variables used in expansions as "used".  This
+  // MUST occur before parameter extraction - otherwise parameters will get wrong values.
+  ExpandWalker exw(_input_filename);
+  _root->walk(&exw);
+  for (auto & var : exw.used)
+    _extracted_vars.insert(var);
+  for (auto & msg : exw.errors)
+    _errmsg += msg + "\n";
+
+  // do as much error checking as early as possible so that errors are more useful instead
+  // of surprising and disconnected from what caused them.
+  DupParamWalker dw(_input_filename);
+  BadActiveWalker bw(_input_filename);
+  _root->walk(&dw, hit::NodeType::Field);
+  _root->walk(&bw, hit::NodeType::Section);
+  for (auto & msg : dw.errors)
+    _errmsg += msg + "\n";
+  for (auto & msg : bw.errors)
+    _errmsg += msg + "\n";
+
+  if (_errmsg.size() > 0)
+    mooseError(_errmsg);
+
+  // There are a few order dependent actions that have to be built first in
+  // order for the parser and application to function properly:
+  //
+  // SetupDebugAction: This action can contain an option for monitoring the parser progress. It must
+  // be parsed first
+  //                   to capture all of the parsing output.
+  //
+  // GlobalParamsAction: This action is checked during the parameter extraction routines of all
+  // subsequent blocks.
+  //                     It must be parsed early since it must exist during subsequent parameter
+  // extraction.
+  //
+  // DynamicObjectRegistration: This action must be built before any MooseObjectActions are built.
+  // This is because
+  //                            we retrieve valid parameters from the Factory during parse time.
+  // Objects must
+  //                            be registered before validParameters can be retrieved.
+  _secs_need_first = {
+      _syntax.getSyntaxByAction("SetupDebugAction", "setup_debug"),
+      _syntax.getSyntaxByAction("GlobalParamsAction", "set_global_params"),
+      _syntax.getSyntaxByAction("DynamicObjectRegistrationAction", "dynamic_object_registration"),
+  };
+
+  // walk all the sections extracting paramters from each into InputParameters objects
+  for (auto & sec : _secs_need_first)
+  {
+    auto n = _root->find(sec);
+    if (n != nullptr)
```

Comment:
I'm not sure about your personal preference, but you don't need to feel obligated to explicitly test against `nullptr`.